npeps/npep-133-fqdn-egress-selector.md

```diff
@@ -0,0 +1,438 @@
+* This enhancement does not require any mechanism for securing DNS resolution
+  (e.g. DNSSEC or DNS-over-TLS). Unsecured DNS requests are expected to be
+  sufficient for looking up FQDNs.
+
+## Introduction
+
+FQDN-based egress controls are a common enterprise security practice.
+Administrators often prefer to write security policies using DNS names such as
+“www.kubernetes.io” instead of capturing all the IP addresses the DNS name might
+resolve to. Keeping up with changing IP addresses is a maintenance burden, and
+hampers the readability of the network policies.
+
+## User Stories
+
+* As a cluster admin, I want to allow all Pods in the cluster to send traffic to
+  an external service specified by a well-known domain name. For example, all
+  Pods must be able to talk to `my-service.com`.
+
+* As a cluster admin, I want to allow Pods in the "monitoring" namespace to be
+  able to send traffic to a logs-sink, hosted at `logs-storage.com`
+
+* As a cluster admin, I want to allow all Pods in the cluster to send traffic to
+  any of the managed services provided by my Cloud Provider. Since the cloud
+  provider has a well known parent domain, I want to allow Pods to send traffic
+  to all sub-domains using a wild-card selector -- `*.my-cloud-provider.com`
+
+### Future User Stories
+
+These are some user stories we want to keep in mind, but due to limitations of
+the existing Network Policy API, cannot be implemented currently. The design
+goal in this case is to ensure we do not make these unimplementable down the
+line.
+
+* As a cluster admin, I want to block all cluster egress traffic by default, and
+  require namespace admins to create NetworkPolicies explicitly allowing egress
+  to the domains they need to talk to.
+
+  The Cluster admin would use a `BaselineAdminNetworkPolicy` object to switch
+  the default disposition of the cluster. Namespace admins would then use a FQDN
+  selector in the Kubernetes `NetworkPolicy` objects to allow `my-service.com`.
+  
+## API
+
+This NPEP proposes adding a new type of `AdminNetworkPolicyEgressPeer` called
+`FQDNPeerSelector` which allows specifying domain names.
+
+```golang
+type AdminNetworkPolicyEgressPeer struct {
+    <snipped>
+    // FQDN provides a way to specify domain names as peers.
+    //
+    // FQDN is only supported for ALLOW rules. In order to control access, FQDN
+    // ALLOW rules should be used with a lower priority egress deny -- this
+    // allows the admin to maintain an explicit "allowlist" of reachable
+   // domains.
+    //
+    // Support: Extended
+    //
+    // +optional
+    FQDN *FQDNPeerSelector `json:"fqdn,omitempty"`
+}
+
+type FQDNPeerSelector struct {
+    // Name specifies the exact Fully-Qualified Domain Name to select (the
+    // trailing '.' is optional)
+    //
+    // Support: Extended
+    //
+    // +optional
+    // +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]([-a-zA-Z0-9_]*[a-zA-Z0-9])*\.?)*$`
+    Name string `json:"names,omitempty"`
```

Comment:
is this not a subset of Pattern?

Reply:
It is. The reason I split them is we had discussions originally about core vs extended functionality. There was a concern from folks that some NetPol implementations couldn't support wildcards, but could support a direct match.

Splitting the fields gives those implementations a good way to indicate what they support (they can reject policies which set `pattern`). I'm not sure if we still want that approach

@astoycos do you have thoughts? @tssurya as well since I think you did some work with core vs extended definitions

Reply:
If we want to have separate fields for wildcard and non-wildcard matches sounds like a good section for the NPEP. Let's discuss it during the next meeting and write it down for future reference

Reply:
it feels weird honestly having both, and this make necessary to treat this as an union, complicating validations, both can not be set at same time

Reply:
Calico's policy model only has one field and it feels quite natural in use so that'd be our preference.

Reply:
We discussed this last meeting, and the plan of record is to only support pattern, not name.